flink-staging/flink-scala-shell/src/main/scala/org/apache/flink/api/scala/FlinkShell.scala

```diff
@@ -0,0 +1,148 @@
+        opt[(String)] ("addclasspath") abbr("a") valueName("<path/to/jar>") action {
+          case (x,c) =>
+            val xArray = x.split(":")
+            c.copy(externalJars = Option(xArray))
+          } text("specifies additional jars to be used in Flink\n")
+        )
+
+      cmd("remote") action { (_, c) =>
+        c.copy(flinkShellExecutionMode = REMOTE)
+      } text("starts Flink scala shell connecting to a remote cluster\n") children(
+        arg[String]("<host>") action { (h, c) =>
+          c.copy(host = h) }
+          text("remote host name as string"),
+        arg[Int]("<port>") action { (p, c) =>
+          c.copy(port = p) }
+          text("remote port as integer\n"),
+        opt[(String)] ("addclasspath") abbr("a")  valueName("<path/to/jar>") action {
+          case (x,c) =>
+            val xArray = x.split(":")
+            c.copy(externalJars = Option(xArray))
+          } text("specifies additional jars to be used in Flink")
+
+      )
+      help("help") abbr("h") text("prints this usage text\n")
+    }
+
+
+    // parse arguments
+    parser.parse (args, Config () ) match {
+      case Some(config) =>
+        startShell(config.host,
+          config.port,
+          config.flinkShellExecutionMode,
+          config.externalJars)
+
+      case _ => println("Could not parse program arguments")
+    }
+  }
+
+
+  def startShell(
+      userHost : String, 
+      userPort : Int,
+      executionMode : Int,
+      externalJars : Option[Array[String]] = None): Unit ={
+    
+    println("Starting Flink Shell:")
+
+    var cluster: LocalFlinkMiniCluster = null
+
+    // either port or userhost not specified by user, create new minicluster
+    val (host,port) = if (executionMode == LOCAL) {
+      cluster = new LocalFlinkMiniCluster(new Configuration, false)
+      cluster.start()
+      val port = cluster.getLeaderRPCPort
+      println(s"\nStarting local Flink cluster (host: localhost, port: $port).\n")
+      ("localhost",port)
+    } else if(executionMode == UNDEFINED) {
+      println("Error: please specify execution mode:")
+      println("[local | remote <host> <port>]")
+      return
+    } else if(userHost == "none" || userPort == -1){
+      println("Error: <host> or <port> not specified!")
+      return
+    } else {
+      println(s"\nConnecting to Flink cluster (host: $userHost, port: $userPort).\n")
+      (userHost, userPort)
+    }
+
+    
+    // custom shell
+    var repl : FlinkILoop = null;
+    if(bufferedReader == null) {
+      repl = new FlinkILoop(host, port, externalJars) //new MyILoop();
+    } else {
+      val out = new StringWriter()
+      repl = new FlinkILoop(host, port, externalJars,bufferedReader, new  JPrintWriter(out));
+    }
+    repl.settings = new Settings()
+
+    repl.settings.usejavacp.value = true
+
+    // start scala interpreter shell
+    repl.process(repl.settings)
+
+    //repl.closeInterpreter()
+
+    if (cluster != null) {
```

Comment:
Can we put this in a `finally` block to make sure that it's always closed if the option value is defined.